### PR TITLE
[Regex] Fix replace_with() signature

### PR DESCRIPTION
**Pull request by caugner**
The $matches array that is passed to `preg_replace_callback()`'s $callback 
can have string keys when the pattern uses named capture groups.

See: https://github.com/vimeo/psalm/pull/5608

### CODE REVIEW

File: src/Psl/Regex/replace_with.php [hunk 1]
```diff
@@ -15,7 +15,7 @@
  * `$callback`.
  *
  * @param non-empty-string $pattern The pattern to search for.
- * @param (callable(array<int, string>): string) $callback The replacement callable.
+ * @param (callable(string[]): string) $callback The replacement callable.
```
Comment (azjezz) [inline src/Psl/Regex/replace_with.php:18]: same as https://github.com/vimeo/psalm/pull/5608/files#r610848313

**Review by azjezz — CHANGES_REQUESTED** [review]

File: src/Psl/Regex/replace_with.php (hunk 1 above)
Comment (azjezz) [inline src/Psl/Regex/replace_with.php]:
```suggestion
 * @param (callable(array<array-key, string>): string) $callback The replacement callable.
```

**Review by azjezz — CHANGES_REQUESTED** [review]

File: src/Psl/Regex/replace_with.php (hunk 1 above)
Comment (azjezz) [reply src/Psl/Regex/replace_with.php]: we don't use phpstorm generics here :)

**Comment by azjezz** [comment]
could you please add some unit tests and static analysis tests for this?

File: src/Psl/Regex/replace_with.php (hunk 1 above)
Comment (caugner) [reply src/Psl/Regex/replace_with.php]: `string[]` and `array<array-key, string>` are exactly the same: https://psalm.dev/r/212f934d6c

Do you still prefer the more verbose version?

Comment (caugner) [reply src/Psl/Regex/replace_with.php]: `string[]` is actually a valid array type according to the proposed PSR-5 PHPDoc: https://github.com/php-fig/fig-standards/blob/d82fd79a13ab3cd2676ed65dafbf4196de9fd322/proposed/phpdoc.md#abnf

**Comment by caugner** [comment]
> could you please add some unit tests and static analysis tests for this?

Since I didn't actually fix any code, just PHPDoc, I'm not sure what could/should be unit tested here. Once https://github.com/vimeo/psalm/pull/5608 is merged and you upgrade to the next Psalm version, your code will produce an error *without this change* though.

PS: The main reason why I created this PR is that `vimeo/psalm` tests a fork of `psl` as part of CI.

File: src/Psl/Regex/replace_with.php (hunk 1 above)
Comment (azjezz) [reply src/Psl/Regex/replace_with.php]: yes, we don't use `T[]` syntax in PSL

**Review by azjezz — APPROVED** [review]

**Pull request merged**